ionc/ion_timestamp.c

```diff
@@ -1305,6 +1308,9 @@ iERR ion_timestamp_binary_read(ION_STREAM *stream, int32_t len, decContext *cont
     // now we set the "has timezone" bit
     if (has_offset) {
         SET_FLAG_ON(ptime->precision, ION_TT_BIT_TZ);
+        ptime->tz_offset = (int16_t)-offset;
+        IONCHECK(_ion_timestamp_to_utc(ptime, ptime));
+        ptime->tz_offset = (int16_t)offset;
```

Comment:
Because of how timestamps are stored in Ion binary. All of their fields are stored in UTC, but the offset field still refers to the local offset from UTC. So when binary timestamps are read, they need to be converted from UTC to local time, with the offset preserved. The simplest way to do this is to negate the offset, perform a UTC conversion on the resulting timestamp, and then restore the correct offset. That's what this does.